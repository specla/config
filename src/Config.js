import dot from '@specla/dot-string'

export default class Config {
  /**
   * Create a new configuration instance
   * @param  {Object} config
   * @return {Config}
   */
  constructor (config = {}) {
    if (typeof config !== 'object' || Array.isArray(config)) {
      throw new Error(`Config does only accept an object as the first argument`)
    }

    this._flattenConfig = {}
    this._config = {}
    this._parse(config)
  }

  /**
   * Parse the config object and create a flatten config
   * @param  {Object} config
   * @return {Object}
   */
  _parse (config) {
    this._flatten(config)

    for (const key in this._flattenConfig) {
      this.set(key, this._flattenConfig[key])
    }
  }

  /**
   * Flatten objects into an object of dot strings
   * @param  {Mixed} config
   * @param  {Array} context
   * @return {Mixed}
   */
  _flatten (config, context = []) {
    if (!Array.isArray(config) && typeof config === 'object') {
      return this._flattenObject(config, context)
    }

    return (this._flattenConfig[context.join('.')] = config)
  }

  /**
   * Flatten object
   * @param  {Mixed} config
   * @param  {Array} context
   */
  _flattenObject (config, context) {
    for (const key in config) {
      this._flatten(config[key], context.concat(key))
    }
  }

  /**
   * Loop through the config object an invoke functional properties
   * @param  {Object} config
   * @return {Object}
   */
  _parseConfig (config) {
    const copyConfig = Object.assign({}, config)

    for (const key in copyConfig) {
      if (typeof copyConfig[key] === 'function') {
        copyConfig[key] = copyConfig[key](this)
      }

      if (!Array.isArray(copyConfig[key]) && typeof copyConfig[key] === 'object') {
        copyConfig[key] = this._parseConfig(copyConfig[key])
      }
    }

    return copyConfig
  }

  /**
   * Get config property from dot notation
   * If the property is a function then invoke it with the config object
   * as the first argument
   * @param  {String} key
   * @return {Mixed}
   */
  get (key) {
    let value

    if (!key) {
      value = this._parseConfig(this._config)
    }

    if (!value && this._flattenConfig[key]) {
      value = this._flattenConfig[key]
    }

    if (!value) {
      try {
        value = dot.get(this._config, key)
      } catch (err) {}
    }

    if (!Array.isArray(value) && typeof value === 'object') {
      return this._parseConfig(value)
    }

    if (typeof value === 'function') {
      return value(this)
    }

    return value
  }

  /**
   * Set value for a given key in dot notation
   * @param {String} key
   * @param {Mixed} value
   * @return {Mixed}
   */
  set (key, value) {
    if (typeof value !== 'object' || Array.isArray(value)) {
      this._flattenConfig[key] = value
      return dot.set(this._config, key, value)
    }

    for (const prop in value) {
      this.set(`${key}.${prop}`, value[prop])
    }
  }

  /**
   * Delete a config property with dot notation
   * @param {String} key
   */
  unset (key) {
    if (!key.includes('.')) {
      key += '.'
    }

    for (const prop in this._flattenConfig) {
      if (prop.includes(`${key}`)) {
        delete this._flattenConfig[prop]
      }
    }

    this._config = {}
    this._parse(this._flattenConfig)
  }

  /**
   * Merge a new config with the current config
   * @param  {Object} config
   * @return {Config}
   */
  merge (config) {
    if (config instanceof Config) {
      config = this._flattenConfig
    }

    this._parse(Object.assign({}, this._flattenConfig, config))

    return this
  }
}
