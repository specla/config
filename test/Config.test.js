/* eslint-env jest */
import Config from 'Config'

test('Should create a config tree structure and a flatten structure', () => {
  const config = new Config({
    'hello.key': true,
    hello: {
      world: 325
    },
    array: ['test']
  })
  expect(config._config).toMatchSnapshot()
  expect(config._flattenConfig).toMatchSnapshot()
})

test('Should throw error if invalid arguments are given to the constructor', () => {
  expect(() => new Config('test')).toThrowErrorMatchingSnapshot()
  expect(() => new Config([])).toThrowErrorMatchingSnapshot()
})

test('Should access config property with dot notation', () => {
  const config = new Config({
    'key.value': true
  })

  expect(config._config.key.value).toBe(true)
  expect(config._flattenConfig['key.value']).toBe(true)
  expect(config.get('key.value')).toBe(true)
})

test('Should return object with all config properties', () => {
  const config = new Config({
    'key.value': true,
    hello: 'world',
    'func': config => 'hello ' + config.get('hello'),
    key: {
      'func': config => config.get('hello'),
      testing: false
    }
  })

  expect(config.get()).toMatchSnapshot()
  expect(config.get().key.value).toBe(true)
  expect(config.get('key')).toMatchSnapshot()
})

test(`Should return undefined if a config property isn't found`, () => {
  const config = new Config()
  expect(config.get('hello.world')).toBe(undefined)
})

test('Should mutate a config property with dot notation', () => {
  const config = new Config({
    'key.value': false
  })

  config.set('key.value', true)
  config.set('key', {
    'test.newValue': true
  })

  expect(config._config.key.value).toBe(true)
  expect(config._flattenConfig['key.value']).toBe(true)
  expect(config.get('key.value')).toBe(true)
  expect(config.get('key.test.newValue')).toBe(true)
})

test('Should set override object with new object', () => {
  const config = new Config({
    'key.value': false
  })

  config.set('key', {
    value: true,
    hello: 'world'
  })

  expect(config._config.key.hello).toBe('world')
  expect(config._flattenConfig['key.hello']).toBe('world')
  expect(config.get('key.hello')).toBe('world')
  expect(config._config.key.value).toBe(true)
  expect(config._flattenConfig['key.value']).toBe(true)
  expect(config.get('key.value')).toBe(true)
})

test('Should merge existing config with a new object', () => {
  const config = new Config({ 'test.key': true })

  config.merge({
    'test.newKey': 325,
    hello: 'world'
  })

  expect(config._config).toMatchSnapshot()
  expect(config._flattenConfig).toMatchSnapshot()
})

test('Should delete a config property', () => {
  const config = new Config({
    'my.config': true,
    'some.key': true,
    'some.test': true,
    'someOtherKey': true,
    hello: 'world'
  })

  config.unset('my.config')
  config.unset('some')
  expect(config._config).toMatchSnapshot()
})

test('Should merge existing config with a new config instance', () => {
  const config = new Config({ 'test.key': true })

  const instance = config.merge(new Config({
    'test.newKey': 325,
    hello: 'world'
  }))

  expect(config._config).toMatchSnapshot()
  expect(config._flattenConfig).toMatchSnapshot()
  expect(instance).toBe(config)
})

test('Should invoke the functional property and parse it the config as an argument', () => {
  const config = new Config({
    'test.key': 'test',
    'test.func.prop': config => config.get('test.key') + ' value'
  })

  expect(config.get('test.func.prop')).toBe('test value')
})
