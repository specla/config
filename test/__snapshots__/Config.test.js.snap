// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should create a config tree structure and a flatten structure 1`] = `
Object {
  "array": Array [
    "test",
  ],
  "hello": Object {
    "key": true,
    "world": 325,
  },
}
`;

exports[`Should create a config tree structure and a flatten structure 2`] = `
Object {
  "array": Array [
    "test",
  ],
  "hello.key": true,
  "hello.world": 325,
}
`;

exports[`Should delete a config property 1`] = `
Object {
  "hello": "world",
  "someOtherKey": true,
}
`;

exports[`Should merge existing config with a new config instance 1`] = `
Object {
  "test": Object {
    "key": true,
  },
}
`;

exports[`Should merge existing config with a new config instance 2`] = `
Object {
  "test.key": true,
}
`;

exports[`Should merge existing config with a new object 1`] = `
Object {
  "hello": "world",
  "test": Object {
    "key": true,
    "newKey": 325,
  },
}
`;

exports[`Should merge existing config with a new object 2`] = `
Object {
  "hello": "world",
  "test.key": true,
  "test.newKey": 325,
}
`;

exports[`Should return object with all config properties 1`] = `
Object {
  "func": "hello world",
  "hello": "world",
  "key": Object {
    "func": "world",
    "testing": false,
    "value": true,
  },
}
`;

exports[`Should return object with all config properties 2`] = `
Object {
  "func": "world",
  "testing": false,
  "value": true,
}
`;

exports[`Should throw error if invalid arguments are given to the constructor 1`] = `"Config does only accept an object as the first argument"`;

exports[`Should throw error if invalid arguments are given to the constructor 2`] = `"Config does only accept an object as the first argument"`;
